### PR TITLE
Try harder to use the C compiler for compiling asm

### LINKED ISSUE #1208: Ability to configure project without a language
The use case would be something like a Gnome application written purely in a scripting language such as Python. As such it never needs a C compiler to build any libraries but it is still far more useful than setuptools because you have easy access to building GResource files, configuring desktop files, handling translations, and it isn't setuptools which is awful.

### LINKED ISSUE #1370: Compilation/linking of executable with ASM source uses C++
As documented in:
https://gist.github.com/fooishbar/37a6b71b7c028bb1df8f1ea1ff09ba92

That executable contains some .c files and a .S file, added to sources the normal way with a plan executable() target. The .S file is compiled with the C++ compiler, which seems to cause the final executable to be linked with the C++ compiler/linker, and we gain an additional unwanted runtime dependency on libstdc++ as a result.

### PR DESCRIPTION
**Pull request by nirbheek**
Use an ordered dict for the compiler dictionary and sort it according to a priority order: fortran, c, c++, etc. This also ensures that builds are reproducible because it would be a toss-up whether a C or a C++ compiler would be used based on the order in which `compilers.items()` would return items. Closes #1370

Also don't require a language/compiler for configuring. Closes #1208

### CODE REVIEW

**Comment by msink** [comment]
Maybe better decouple asm into separate compiler?
In gcc/clang it is the same as C, but msvc use cl for C and ml/ml64 for asm.

**Comment by nirbheek** [comment]
Perhaps, but that's a lot more work and I don't see the gain. The comparison with `ml` is not the same, because `ml` does not do pre-processing, so you need to pass your code through `cl -E` and then through `ml`. 

We do need to add native support for that somehow, but in the meantime you can do this: 
https://github.com/mesonbuild/meson/blob/master/test%20cases/common/126%20llvm%20ir%20and%20assembly/meson.build#L40
https://github.com/centricular/libffi/blob/meson/src/meson.build#L55

**Comment by jpakkane** [comment]
There does not seem to be a test that defines a target with C, C++ and asm files and checks that the .S file is actually compiled with C.

**Comment by nirbheek** [comment]
True, forgot about that. Will add it!

**Comment by nirbheek** [comment]
PR should be ready to merge now.

**Comment by jpakkane** [comment]
The original bug that daniels reported was that a plain C + asm target was linked with g++ because it interprets `.s` as a C++ file. That is not what this is testing for. There should be a test with a project that has C and C++ enabled, which builds a target with only `.c` and `.s` and then verifies that the linker used is `gcc`, not `g++`. (linkers are sadly not exported to the compile db)

**Comment by nirbheek** [comment]
AFAICT, libstdc++ will only be linked if symbols actually need it, i.e., if the `.S` file was compiled as a C++ file, and the c-cpp-asm test is a superset of the c-asm test. But I'll add tests for those anyway, just for completeness.

File: run_tests.py [hunk 1]
```diff
@@ -32,7 +38,10 @@
         units += ['LinuxlikeTests']
     elif mesonlib.is_windows():
         units += ['WindowsTests']
-    returncode += subprocess.call([sys.executable, 'run_unittests.py', '-v'] + units)
+    # Unit tests always use the Ninja backend, so just skip them if we're
+    # testing the VS backend
+    if not using_vs_backend()
```
Comment (ignatenkobrain) [inline run_tests.py]: __[flake8]__
* [E901] SyntaxError: invalid syntax

File: run_tests.py [hunk 2]
```diff
@@ -32,7 +38,10 @@
...
+        returncode += subprocess.call([sys.executable, 'run_unittests.py', '-v'] + units)
```
Comment (ignatenkobrain) [inline run_tests.py]: __[flake8]__
* [E113] unexpected indentation

**Comment by nirbheek** [comment]
The new commits should do that. Travis is down, so the CI is not going to start up anytime soon.

**Pull request merged**